longhaul/longhaul.md

```diff
@@ -0,0 +1,235 @@
+# IoT SDK Long Haul Technical Spec
+
+The Longhaul tests are to ensure the SDK can handle the rigors of in-field life.
+
+## Delivery scheduled
+
+* Longhaul tests will follow a crawl-\>walk-\>run approach.  
+
+* onghaul tests will sim-ship tests for both node and python using the Horton framework.
+
+## Parameters for Longhaul Testing
+
+* The test should run between 12 and 336 hours (2 weeks), but could be run for much longer periods
+
+* The test will perform a variety of operations and verify the results.  For example, D2C telemetry test will verify that the telemetry is sent successfully and also  arrived at the EventHub
+
+### Crawl
+
+* Tests will send D2C messages at a fixed cadence and verify receipt
+
+* Tests will use those same D2C to deliver time-series telemetry (such as memory use, ) to the hub for dashboarding
+
+* Tests will use reported properties to report test configuration and current status to the hub (elapsed time, average latency, etc)
+
+* Tests will run until failure, 
+
+* Failure conditions will include crash, timeout on message receive, timeout on send operation, and heap level checks.
+
+* Tests will be manually launched and manually verified.
+
+### Walk
+
+* Tests will use more iothub features (twin, methods, C2D, etc).
+
+* Fault injection tests will operate the same as simple D2C tests, but will launch as a seperate test scenario.
+
+* Simple randomization of these operations will be supported.  Overlapping of different operations won't need to happen here.
+
+* test launch automation will be increased so tests can be run in Docker containers in more of a bulk fashion
+
+
+### Run
+
+* More test scenarios will be supported, hopefully with the same test core running with different options.
+
+* Tests can be controlled via desired properties and C2D messages.  This includes operation cadence, failure properties, fault rates, and operation choices.
+
+* Maximum chaos is supported.  
+
+* DPS is used to provision devices to test
+
+* Tests are run with all manner of authenticatoin (x509, symmetric key, edge, etc)
+
+## Operations to test
+
+Operations will be phased in as follows.  These operation names are also used inside the various test structures below.
+
+| Operation        | Stage |
+|------------------|-------|
+| D2C              | crawl |
+| C2D              | walk  |
+| getTwin          | walk  |
+| updateTwin       | walk  |
+| receiveTwinPatch | walk  |
+| handleMethod     | walk  |
+| invokeMethod     | walk  |
+| DPS register     | run   |
+| blob upload      | run   |
+
+
+## Lonhaul configuration and progress
+
+* D2C telemetry will be used to send time-series data for deshboarding.
```

Comment:
any ideas how this telemetry will be reported? something I'm excited to dig more into because having a dashboard for these errors and key stats for each test run will be essential